src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -451,27 +451,43 @@ private async Task AddHttp11ConnectionAsync(RequestQueue<HttpConnection>.QueueIt
         {
             if (NetEventSource.Log.IsEnabled()) Trace("Creating new HTTP/1.1 connection for pool.");
 
-            HttpConnection connection;
-            using (CancellationTokenSource cts = GetConnectTimeoutCancellationTokenSource())
+            HttpConnectionWaiter<HttpConnection> waiter = queueItem.Waiter;
+            HttpConnection? connection = null;
+            Exception? connectionException = null;
+
+            CancellationTokenSource cts = GetConnectTimeoutCancellationTokenSource();
+            waiter.ConnectionCancellationTokenSource = cts;
+            try
             {
-                try
-                {
-                    connection = await CreateHttp11ConnectionAsync(queueItem.Request, true, cts.Token).ConfigureAwait(false);
-                }
-                catch (OperationCanceledException oce) when (oce.CancellationToken == cts.Token)
-                {
-                    HandleHttp11ConnectionFailure(queueItem.Waiter, CreateConnectTimeoutException(oce));
-                    return;
-                }
-                catch (Exception e)
+                connection = await CreateHttp11ConnectionAsync(queueItem.Request, true, cts.Token).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException oce) when (oce.CancellationToken == cts.Token)
+            {
+                connectionException = CreateConnectTimeoutException(oce);
+            }
+            catch (Exception e)
+            {
+                connectionException = e;
+            }
```

Comment:
I considered it but to me it seemed less readable. Do we generally prefer shorter forms over more expressive ones?

Reply:
I personally don't think multiple catch blocks is "more expressive".  And for me it's not about length, it's about understandability: we get an exception and we need to store either that exception or one that wraps it, which we can express concisely and readably with a single catch block and ternary, making it clear exactly what the difference in behavior is based on the kind of exception, rather than two catch blocks that need to be compared to understand what the difference between them is.